### PR TITLE
feat: [LI-181] implement ui for candidate select interview time

### PR DESCRIPTION
**Pull request by lulu0729**
### Issue ticket number and link
[LI-181](https://predictivehire.atlassian.net/browse/LI-181)
### Figma Screenshot
image

### Describe main changes

https://github.com/PredictiveHire/phapp-schedule-link/assets/34208317/9c9032cc-70fa-4003-b633-469095b97278
- implement ui for candidate select interview time
- add unit tests for utils and components

### Checklist in Local 
- [x] I have extract component business logic into custom hook
- [x] I have added unit tests
- [ ] I have integrated with backend api and done a e2e test locally
- [ ] I have updated the graphql query types

### Checklist in Dev
- [ ] I have deployed to dev
- [ ] I have integrated with backend api and done a e2e test

### Checklist before requesting a review
- [x] I have done the **code format** using `pnpm lint:fix` and `pnpm format:fix`
- [x] I have **rebase squash** current branch
- [ ] I have confirmed **Naming Convention** with team if needed.
- [ ] I have done a **self review** in GitHub.

### Checklist in Qa Environment
- [ ] I have deployed to qa and let QA team do a test
- [ ] I have fix the QA bugs after QA team finish testing
- [ ] I have put the screenshot for Qa environment integration test below:

### Checklist in Sandbox Environment
- [ ] I have deployed to sandbox and let QA team do test
- [ ] I have fix the Sandbox Bugs after QA team finish testing
- [ ] I have put the screenshot for Qa environment integration test below:

### Checklist after Merge
- [ ] I have known the dependencies of release deployment of current PR.
- [ ] I have reviewed the release pipe from team lead to secure the production deployment.
- [ ] I have update the Jira tickets status after all task has been finished.

[LI-181]: https://predictivehire.atlassian.net/browse/LI-181?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by yingzhi-lu — COMMENTED** [review]
Hi @lulu0729 looks like we did not calling branding API to get customer based branding setting, did we?

File: src/pages/schedule-interview/__test__/ScheduleInterviewPage.test.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { ScheduleInterviewPage } from "../ScheduleInterviewPage"
+
+// Mock the MobileCandidateSelectInterviewTime component
+jest.mock("../components/MobileCandidateSelectInterviewTime", () => {
+  return {
+    MobileCandidateSelectInterviewTime: () => (
+      <div data-testid="mock-mobile-candidate-select-interview-time">Mock MobileCandidateSelectInterviewTime</div>
+    ),
+  }
+})
```
Comment (martin-luo) [inline src/pages/schedule-interview/__test__/ScheduleInterviewPage.test.tsx:13]: why do we mock the child component here?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    })
+
+  const disabledDate = (date: Dayjs) => !isInterviewDate(date)
+
+  const clickDate = (date: Dayjs) => {
+    if (isInterviewDate(date)) {
+      setInterviewDate(date)
+      setShowTimeSlotBooking(true)
+    }
+  }
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx:51]: should we extract these logics into a hook?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                  onClick={() => onChange(value.clone().add(1, "month"))}
+                  className="text-secondary cursor-pointer text-xl"
+                />
+              </div>
+            )}
+            fullCellRender={fullCellRender}
+            disabledDate={disabledDate}
+          />
+        </div>
+      )}
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx:95]: too long, better to break into smaller functions like `funnCellRender`

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/__test__/MobileCandidateSelectInterviewTime.test.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+jest.mock("../components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking", () => ({
+  MobileInterviewTimeSlotBooking: ({ closeBooking, interviewTimes, initialDate }: InterviewTimeSlotBookingProps) => (
+    <div>
+      Mocked MobileInterviewTimeSlotBooking
+      <button onClick={closeBooking}>Close</button>
+      <div>Initial Date: {initialDate.format("YYYY-MM-DD")}</div>
+      <div>Interview Times: {JSON.stringify(interviewTimes)}</div>
+    </div>
+  ),
+}))
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/__test__/MobileCandidateSelectInterviewTime.test.tsx:23]: same question here, why do we want to mock the child component?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/__test__/MobileCandidateSelectInterviewTime.test.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    await userEvent.click(dateCell)
+
+    expect(screen.getByText("Mocked MobileInterviewTimeSlotBooking")).toBeInTheDocument()
+  })
+
+  it("should closes MobileInterviewTimeSlotBooking when close button is clicked", async () => {
+    render(<MobileCandidateSelectInterviewTime />)
+
+    const dateCell = screen.getAllByText("1")[0] // Assuming the first date cell with "1" is valid
+    await userEvent.click(dateCell)
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/__test__/MobileCandidateSelectInterviewTime.test.tsx:50]: this is not the recommended way to mock user action any more https://testing-library.com/docs/user-event/setup

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewInfo/MobileInterviewInfo.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { LIInterviewMode, LIInterviewModeLabel } from "../../constants"
+
+interface MobileInterviewInfoProps {
+  interviewInfo: {
+    logo: string
+    jobRequisitionName: string
+    interviewMode: string
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewInfo/MobileInterviewInfo.tsx:7]: why not make the type `LIInterviewMode` directly?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        <Button
+          type="text"
+          icon={<RightOutlined />}
+          onClick={() => handleDateChange("next")}
+          disabled={isRightArrowDisabled}
+        />
+      </div>
+      <Radio.Group className="w-full" onChange={(e) => handleTimeChange(e.target.value as string)} value={selectedTime}>
+        {timeSlots.map((slot, index) => (
+          <Radio.Button key={index} value={slot} className="w-full px-5 text-left text-black">
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx:68]: index is not the best way to use as `key`, can we use something else like `slot`?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            {selectedDateAndTime.selectedTime === slot && selectedDateAndTime.selectedDate === currentDate && (
+              <CheckCircleFilled className="checkCircleFilled mr-3 rounded-full bg-[#FFCEFF] text-black" />
+            )}
+            {slot}
+          </Radio.Button>
+        ))}
+      </Radio.Group>
+      <button
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx:76]: why we use `button` and `Button` in one component at the same time?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        className={`buttonBook mt-4 w-full text-base font-normal ${selectedDateAndTime.selectedTime ? "active" : ""}`}
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/MobileInterviewTimeSlotBooking/MobileInterviewTimeSlotBooking.tsx:77]: why not use `cn`?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/__test__/MobileInterviewInfo.test.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    expect(getByAltText("Logo")).toHaveAttribute("src", interviewInfo.logo)
+    expect(getByText(interviewInfo.jobRequisitionName)).toBeInTheDocument()
+    expect(getByText("Interview Type")).toBeInTheDocument()
+    expect(getByText(LIInterviewModeLabel[interviewInfo.interviewMode])).toBeInTheDocument()
+  })
+
+  it("should renders the interview address if the interview mode is IN_PERSON", () => {
+    const { getByText } = render(<MobileInterviewInfo interviewInfo={interviewInfo} />)
+
+    if (interviewInfo.interviewMode === LIInterviewMode.IN_PERSON) {
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/__test__/MobileInterviewInfo.test.tsx:28]: this does not make sense, we should not have this kind of if statement in the unit test

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/__test__/MobileInterviewInfo.test.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+  it("should renders the interview link if the interview mode is ONLINE", () => {
+    const onlineInterviewInfo = {
+      ...interviewInfo,
+      interviewMode: LIInterviewMode.ONLINE,
+    }
+
+    const { getByText } = render(<MobileInterviewInfo interviewInfo={onlineInterviewInfo} />)
+
+    if (onlineInterviewInfo.interviewMode === LIInterviewMode.ONLINE) {
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/components/__test__/MobileInterviewInfo.test.tsx:42]: same here

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useTimeSlots.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        .filter((slot) => slot.start.startsWith(currentDate))
+        .map(
+          (slot) =>
+            `${new Date(slot.start).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" })} - ${new Date(slot.end).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" })}`
+        )
+      setTimeSlots(slots)
+    }
+    updateTimeSlots()
+  }, [currentDate, interviewTimes])
+
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useTimeSlots.ts:18]: why do we need to use `useState` and `useEffect` hook here? I think we can define `updateTimeSlots` as a function to take both `currentDate` and `interviewTimes` as parameter, and export this function instead?

File: src/utils/__test__/regionMapping.test.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    AP_SOUTHEAST_2: "ap-southeast-2",
+    EU_WEST_1: "eu-west-1",
+    US_EAST_1: "us-east-1",
+  },
+  AWSShortRegion: {
+    APSE2: "apse2",
+    EUW1: "euw1",
+    USE1: "use1",
+  },
+}))
```
Comment (martin-luo) [inline src/utils/__test__/regionMapping.test.ts:16]: why do we need to mock constants?

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import { Dayjs } from "dayjs"
+import { useState } from "react"
+
+export const useInterviewBooking = (initialDate: Dayjs, availableDates: string[]) => {
+  const [currentDate, setCurrentDate] = useState<string>(initialDate.format("YYYY-MM-DD"))
+  const [selectedDateAndTime, setSelectedDateAndTime] = useState<{ selectedDate: string; selectedTime: string }>({
+    selectedDate: "",
+    selectedTime: "",
+  })
+  const [selectedTime, setSelectedTime] = useState<string | null>(null)
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts:10]: why do we need `selectedTime` when this value is already captured in `selectedDateAndTime.selectedTime`?

**Review by martin-luo — COMMENTED** [review]
I have some comments, please have a look

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts (hunk 14 above)
Comment (lulu0729) [reply src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts:10]: When switching dates, the displayed time slots will also change. The `selectedTime` refers to the selected time within the currently displayed time slots. The `selectedDateAndTime` is a record of the chosen date and time, which may not be within the currently displayed time slots.

**Comment by lulu0729** [comment]
> Hi @lulu0729 looks like we did not calling branding API to get customer based branding setting, did we?

yes, we may need to implement it in another ticket

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts (hunk 14 above)
Comment (martin-luo) [reply src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useInterviewBooking.ts:22]: great, thanks for the explanation @lulu0729

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useTimeSlots.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    const slots = formatTimeSlots(interviewTimes, currentDate)
+
+    setTimeSlots(slots)
+  }
+
+  return {
+    updateTimeSlots,
+    timeSlots,
+  }
+}
```
Comment (martin-luo) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useTimeSlots.ts:25]: I am not sure why we need to store `timeSlots` as `state` here still. We only use the hook in `MobileInterviewTimeSlotBooking`, and we only used this hook to transform the interviewTimes to timeslots right?

```suggestion
export const useTimeSlots = () => {
  const formatTimeSlots = (interviewTimes: InterviewTimeSlot[], currentDate: string): string[] =>
  interviewTimes
    .filter((slot) => slot.start.startsWith(currentDate))
    .map(
      (slot) =>
        `${new Date(slot.start).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" })} - ${new Date(slot.end).toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" })}`
    )

  return { formatTimeSlots }
}
```

And in `MobileInterviewTimeSlotBooking`, we only need to 
```
  const { formatTimeSlots }  = useTimeSlots()
  const timeSlots = formatTimeSlots(interviewTimes, currentDate)

```

In this way, the hook actually plays the role of a collection of util functions

Comment (lulu0729) [reply src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/hooks/useTimeSlots.ts:25]: Thanks @martin-luo . I understand what you mean

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+// TODO: Mock data from backend
+const interviewInfo = {
+  logo: logoSvg,
+  jobRequisitionName: "job requisition name",
+  interviewMode: LIInterviewMode.IN_PERSON,
+  interviewAddress: "216 Lower Heidelberg Rd, Ivanhoe East VIC 3079",
+}
+
+export const MobileCandidateSelectInterviewTime = () => {
```
Comment (GeekEast) [inline src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx:23]: why we need to add `Mobile` as prefix.

the component should be responsive, right?

do you use media query anywhere else?

**Review by martin-luo — APPROVED** [review]
Looks good to me, great work!

File: src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx (hunk 16 above)
Comment (GeekEast) [reply src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx:23]: Hi @lulu0729 is there any way we could detect the current device is mobile or desktop and then render the corresponding component, for example, `MobileCandidateSelectInterviewTime`

Comment (lulu0729) [reply src/pages/schedule-interview/components/MobileCandidateSelectInterviewTime/MobileCandidateSelectInterviewTime.tsx:23]: Referencing `phapp-talent-hub`, the current method to detect if it is a desktop is: `import { Grid } from "antd"`

**Review by GeekEast — APPROVED** [review]

**Pull request merged**